connaisseur/keys.py

```diff
@@ -0,0 +1,93 @@
+import base64
+import re
+
+import ecdsa
+import rsa
+
+from connaisseur.exceptions import NoSuchClassError
+
+KMS_REGEX = r"^(awskms|gcpkms|azurekms|hashivault|k8s):\/{2,3}[a-zA-Z0-9_.+\/:-]+$"
+KEYLESS_REGEX = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
+
+
+class KeyInterface:
+    """
+    Interface form which all keys inherit.
+    """
+
+    def __new__(cls, data: object):
+        instance = super(KeyInterface, cls).__new__(cls)
+        instance.__init__(data)
+        return instance
+
+    def __init__(self, data: object) -> None:
+        self.value = data
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class Key(KeyInterface):
+    """
+    Abstract Key class, containing the public key used for verifying the image signatures.
+    """
+
+    value: object
+
+    def __new__(cls, data: str):
+        try:
+            key_cls, key_data = Key.__get_key_type_cls(data)
+            return key_cls.__new__(key_cls, key_data)
+        except Exception as err:
+            msg = "Cannot discern key type."
+            raise NoSuchClassError(message=msg, key_type=key_cls) from err
```

Comment:
Would key_cls not always be expected to be None? If the misnamed function returns class and data, instantiation should always work, shouldn't it? And in that case, discerning the type wouldn't be the problem, but rather instantiation of the key object or am I missing sth?

Reply:
The instantiation will still fail, since `None.__new__(None, data)` will raise a `TypeError`.

Reply:
Putting the key data there is dangerous as we may have symmetric secrets as root at some point and then this logs a secret. If you want to notify about which key is "bad", log the name from the helm.yml comfig